alica_tracing/include/tracing/AlicaTrace.h

```diff
@@ -0,0 +1,197 @@
+    // as a result, traces should become no-op.
+    RawTracePtr _rawTrace;
+};
+
+using TracePtr = std::unique_ptr<Trace>;
+
+template <class TraceType>
+class TraceFactory : public alica::IAlicaTraceFactory
+{
+public:
+    TraceFactory() = default;
+
+    TraceFactory(const TraceFactory&) = delete;
+    TraceFactory(TraceFactory&&) = delete;
+    TraceFactory& operator=(const TraceFactory&) = delete;
+    TraceFactory&& operator=(TraceFactory&&) = delete;
+
+    ~TraceFactory()
+    {
+        ROS_INFO_STREAM_NAMED(__func__, "Terminating tracing for service " << _serviceName);
+        // allow termination to propogate,
+        std::this_thread::sleep_for(std::chrono::milliseconds(50));
+        opentracing::Tracer::Global()->Close();
+    };
+
+    // Initialize the trace factory once per process
+    // serviceName is the component running the instance, ex: lbc, gbc
+    // configFilePath is the place factory will look out for tracing configuration.
+    // returns false if initialization failed, true otherwise
+    // Will throw if the initialization failed
+    void init(const std::string& serviceName, const std::string& configFilePath)
+    {
+        if (_initialized) {
+            ROS_WARN_STREAM_NAMED(__func__, "Trace factory already initialized.");
+            return;
+        }
+
+        ROS_INFO_STREAM_NAMED(__func__, "Initializing tracing for service " << serviceName);
+        _serviceName = serviceName;
+        try {
+            auto configYAML = YAML::LoadFile(configFilePath);
+            auto config = jaegertracing::Config::parse(configYAML);
+            auto tracer = jaegertracing::Tracer::make(_serviceName, config, jaegertracing::logging::consoleLogger());
+            opentracing::Tracer::InitGlobal(std::static_pointer_cast<opentracing::Tracer>(tracer));
+        } catch (std::exception& e) {
+            ROS_ERROR_STREAM_NAMED(__func__, "Failed to initialize jaeger: " << e.what());
+            throw e;
+        }
+        _initialized = true;
+        ROS_INFO_STREAM_NAMED(__func__, "Initialized tracing for service " << _serviceName);
+    }
+
+    // Add default tags that you want to be reported as part of all the traces generating from this process.
+    // ex: setDefaultTag("agent_id", 1002)
+    void setDefaultTag(const std::string& key, const RawTraceValue& value);
+
+    std::unique_ptr<alica::IAlicaTrace> create(const std::string& opName, std::optional<const std::string> parent = std::nullopt) const
+    {
+        assert(!opName.empty());
+
+        if (!_initialized) {
+            return std::make_unique<TraceType>();
+        }
+
+        std::unique_ptr<alica::IAlicaTrace> trace = std::make_unique<TraceType>(opName, parent);
+        for (const auto& defaultTag : _defaultTags) {
+            trace->setTag(defaultTag.first, defaultTag.second);
+        }
+
+        return trace;
+    }
+
+private:
+    bool _initialized = false;
+    std::unordered_map<std::string, std::string> _defaultTags;
```

Comment:
it's ok to store RawTraceValue here. The limitation of strings only applies to IAlicaTrace & we should extend it in the future anyway